bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java

```diff
@@ -982,8 +983,41 @@ public static void win32_destroyUnusedHandles(Display display) {
 	for (Monitor monitor : display.getMonitors()) {
 	    availableZoomLevels.add(DPIUtil.getZoomForAutoscaleProperty(monitor.getZoom()));
 	}
-	for (Resource resource: ((Device) display).resourcesWithZoomSupport) {
+	Set<ResourceReference> resources = ((Device) display).resourcesWithZoomSupport;
+	Iterator<ResourceReference> iterator = resources.iterator();
+
+	while (iterator.hasNext()) {
+		ResourceReference ref = iterator.next();
+		Resource resource = ref.get();
+		if (resource == null) {
+			iterator.remove(); // Clean up dead reference.
+			continue;
+		}
 		resource.destroyHandlesExcept(availableZoomLevels);
 	}
 }
+
+private static class ResourceReference extends WeakReference<Resource> {
+
+	public ResourceReference(Resource referent) {
+		super(referent);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj) return true;
+		if (!(obj instanceof ResourceReference)) return false;
```

Comment:
This way you can avoid the cast two lines later:
```suggestion
		if (!(obj instanceof ResourceReference passedResource)) return false;
```